Add a 'status' field to feature flags denoting their use case

emw-apple opened the pull request:
#### 429cc4c7898637e262eae873994cbec3c7ffcdf6
<pre>
Add a &apos;status&apos; field to feature flags denoting their use case
<a href="https://bugs.webkit.org/show_bug.cgi?id=247926">https://bugs.webkit.org/show_bug.cgi?id=247926</a>
rdar://92112770

Reviewed by Brent Fulgham.

Add enum types representing different semantic labels for feature flags.
Add status fields to our various feature flag types:

- Cross-platform WebKit: API::ExperimentalFeature.status() and API::InternalDebugFeature.status()
- Cocoa WebKit SPI: -[_WKExperimentalFeature status] and -[_WKInternalDebugFeature status]
- WebKitLegacy: -[WebFeature status]

Update GeneratePreferences.rb to read a &quot;status&quot; field from
WebPreferences yamls, and update its templates to pass this through to
the feature instantiation site.

Currently, individual preferences do not declare their own status;
instead, the generator picks an appropriate status based on which file
the preference is declared in. Eventually, we will merge these files into a
single preferences manifest, and use the status field to determine where
a feature gets exposed.

* Source/WTF/Scripts/GeneratePreferences.rb:
* Source/WebKit/Scripts/PreferencesTemplates/WebPreferencesExperimentalFeatures.cpp.erb:
* Source/WebKit/Scripts/PreferencesTemplates/WebPreferencesInternalDebugFeatures.cpp.erb:
* Source/WebKit/UIProcess/API/APIExperimentalFeature.cpp:
(API::ExperimentalFeature::create):
(API::ExperimentalFeature::ExperimentalFeature):
* Source/WebKit/UIProcess/API/APIExperimentalFeature.h:
* Source/WebKit/UIProcess/API/APIFeatureStatus.h: Copied from Source/WebKitLegacy/mac/WebView/WebFeatureInternal.h.
* Source/WebKit/UIProcess/API/APIInternalDebugFeature.cpp:
(API::InternalDebugFeature::create):
(API::InternalDebugFeature::InternalDebugFeature):
* Source/WebKit/UIProcess/API/APIInternalDebugFeature.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKExperimentalFeature.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKExperimentalFeature.mm:
(-[_WKExperimentalFeature status]):
* Source/WebKit/UIProcess/API/Cocoa/_WKInternalDebugFeature.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKInternalDebugFeature.mm:
(-[_WKInternalDebugFeature status]):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKitLegacy/mac/Scripts/PreferencesTemplates/WebPreferencesExperimentalFeatures.mm.erb:
* Source/WebKitLegacy/mac/Scripts/PreferencesTemplates/WebPreferencesInternalFeatures.mm.erb:
* Source/WebKitLegacy/mac/WebView/WebFeature.h:
* Source/WebKitLegacy/mac/WebView/WebFeature.m:
(-[WebFeature initWithKey:preferenceKey:name:status:details:defaultValue:hidden:]):
(-[WebFeature initWithKey:preferenceKey:name:details:defaultValue:hidden:]): Deleted.
* Source/WebKitLegacy/mac/WebView/WebFeatureInternal.h:

Canonical link: <a href="https://commits.webkit.org/257166@main">https://commits.webkit.org/257166@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5dbe895eb70adabf0ffa31849d9bdeeb57ae93b9

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/98058 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/7276 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/31213 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/107521 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/167792 "Built successfully and passed tests") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/7762 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/36041 "Built successfully") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/90664 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/104136 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/103710 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/5836 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/84660 "Built successfully") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/32909 "layout-tests (failure)") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/5/builds/101606 "Passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/87705 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/89419 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/75826 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/88916 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/1252 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/20853 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/84630 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/1217 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/22357 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/28605 "Passed tests") | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4931 "Failed to push commit to Webkit repository") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/6076 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/44808 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/87468 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/2516 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/41762 "Passed tests") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/19623 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

brentfulgham commented on Source/WTF/Scripts/GeneratePreferences.rb:
```diff
@@ -220,6 +229,11 @@ def initializeParsedPreferences(parsedPreferences, requireHumanReadableName)
+        statuses = ["embedder", "unstable", "internal", "developer", "testable", "preview", "stable"]
+        options["status"] ||= defaultStatus
+        if !statuses.include? options["status"]
+          raise "ERROR: #{options["status"]} not one of the known statuses: #{statuses}"
```
Thank you for this logging -- I was worried about typos!

brentfulgham reviewed (approved):
It looks like there is no new input file yet, so I imagine the new ruby code doesn't run as part of the build. So this seems safe to land without affecting other work.

emw-apple committed:
Add a 'status' field to feature flags denoting their use case
https://bugs.webkit.org/show_bug.cgi?id=247926
rdar://92112770

Reviewed by Brent Fulgham.

Add enum types representing different semantic labels for feature flags.
Add status fields to our various feature flag types:

- Cross-platform WebKit: API::ExperimentalFeature.status() and API::InternalDebugFeature.status()
- Cocoa WebKit SPI: -[_WKExperimentalFeature status] and -[_WKInternalDebugFeature status]
- WebKitLegacy: -[WebFeature status]

Update GeneratePreferences.rb to read a "status" field from
WebPreferences yamls, and update its templates to pass this through to
the feature instantiation site.

Currently, individual preferences do not declare their own status;
instead, the generator picks an appropriate status based on which file
the preference is declared in. Eventually, we will merge these files into a
single preferences manifest, and use the status field to determine where
a feature gets exposed.

* Source/WTF/Scripts/GeneratePreferences.rb:
* Source/WebKit/Scripts/PreferencesTemplates/WebPreferencesExperimentalFeatures.cpp.erb:
* Source/WebKit/Scripts/PreferencesTemplates/WebPreferencesInternalDebugFeatures.cpp.erb:
* Source/WebKit/UIProcess/API/APIExperimentalFeature.cpp:
(API::ExperimentalFeature::create):
(API::ExperimentalFeature::ExperimentalFeature):
* Source/WebKit/UIProcess/API/APIExperimentalFeature.h:
* Source/WebKit/UIProcess/API/APIFeatureStatus.h: Copied from Source/WebKitLegacy/mac/WebView/WebFeatureInternal.h.
* Source/WebKit/UIProcess/API/APIInternalDebugFeature.cpp:
(API::InternalDebugFeature::create):
(API::InternalDebugFeature::InternalDebugFeature):
* Source/WebKit/UIProcess/API/APIInternalDebugFeature.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKExperimentalFeature.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKExperimentalFeature.mm:
(-[_WKExperimentalFeature status]):
* Source/WebKit/UIProcess/API/Cocoa/_WKInternalDebugFeature.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKInternalDebugFeature.mm:
(-[_WKInternalDebugFeature status]):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKitLegacy/mac/Scripts/PreferencesTemplates/WebPreferencesExperimentalFeatures.mm.erb:
* Source/WebKitLegacy/mac/Scripts/PreferencesTemplates/WebPreferencesInternalFeatures.mm.erb:
* Source/WebKitLegacy/mac/WebView/WebFeature.h:
* Source/WebKitLegacy/mac/WebView/WebFeature.m:
(-[WebFeature initWithKey:preferenceKey:name:status:details:defaultValue:hidden:]):
(-[WebFeature initWithKey:preferenceKey:name:details:defaultValue:hidden:]): Deleted.
* Source/WebKitLegacy/mac/WebView/WebFeatureInternal.h:

Canonical link: https://commits.webkit.org/257166@main
